Core/include/Acts/Navigation/NextNavigator.hpp

```diff
@@ -0,0 +1,357 @@
+  ///
+  /// @param [in,out] state is the propagation state object
+  /// @param [in] stepper Stepper in use
+  ///
+  /// @return boolean return triggers exit to stepper
+  template <typename propagator_state_t, typename stepper_t>
+  void initialize(propagator_state_t& state, const stepper_t& stepper) const {
+    ACTS_VERBOSE(volInfo(state) << posInfo(state, stepper) << "initialize");
+
+    auto& nState = state.navigation;
+
+    if (nState.currentDetector == nullptr) {
+      ACTS_VERBOSE("assigning detector from the config.");
+      nState.currentDetector = m_cfg.detector;
+    }
+
+    if (nState.currentDetector == nullptr) {
+      ACTS_ERROR("panic: no detector");
+      return;
+    }
+  }
+
+  /// Navigation (re-)initialisation for the target
+  ///
+  /// This is only called a few times every propagation/extrapolation
+  ///
+  /// ---------------------------------------------------------------------
+  ///
+  /// As a straight line estimate can lead you to the wrong destination
+  /// Volume, this will be called at:
+  /// - initialization
+  /// - attempted volume switch
+  /// Target finding by association will not be done again
+  ///
+  /// @tparam propagator_state_t The state type of the propagagor
+  /// @tparam stepper_t The type of stepper used for the propagation
+  ///
+  /// @param [in,out] state is the propagation state object
+  /// @param [in] stepper Stepper in use
+  ///
+  /// boolean return triggers exit to stepper
+  template <typename propagator_state_t, typename stepper_t>
+  void initializeTarget(propagator_state_t& state,
+                        const stepper_t& stepper) const {
+    ACTS_VERBOSE(volInfo(state)
+                 << posInfo(state, stepper) << "initialize target");
+
+    auto& nState = state.navigation;
+
+    if (nState.currentVolume == nullptr) {
+      nState.currentVolume = nState.currentDetector->findDetectorVolume(
+          state.geoContext, nState.position);
+
+      if (nState.currentVolume != nullptr) {
+        ACTS_VERBOSE(volInfo(state) << "switched detector volume");
+      }
+    }
+
+    if (nState.currentVolume == nullptr) {
+      ACTS_ERROR("panic: no current volume");
+      return;
+    }
+
+    nState.currentVolume->updateNavigationState(state.geoContext, nState);
+  }
+
+  template <typename propagator_state_t, typename stepper_t>
+  void fillNavigationState(propagator_state_t& state, const stepper_t& stepper,
+                           NavigationState& nState) const {
+    nState.position = stepper.position(state.stepping);
+    nState.direction = stepper.direction(state.stepping);
+    nState.absMomentum = stepper.momentum(state.stepping);
+    nState.charge = stepper.charge(state.stepping);
+    auto fieldResult = stepper.getField(state.stepping, nState.position);
+    if (!fieldResult.ok()) {
+      ACTS_ERROR("could not read from the magnetic field");
+    }
+    nState.magneticField = *fieldResult;
+  }
+};
+
+}  // namespace Experimental
```

Comment:
I am pretty happy with this first draft, pretty much along the lines what I thought it should do.